nostr/event.py

```diff
@@ -16,41 +18,65 @@ class EventKind(IntEnum):
     DELETE = 5
 
 
-class Event():
-    def __init__(
-            self, 
-            public_key: str, 
-            content: str, 
-            created_at: int = None, 
-            kind: int=EventKind.TEXT_NOTE, 
-            tags: "list[list[str]]"=[], 
-            id: str=None, 
-            signature: str=None) -> None:
-        if not isinstance(content, str):
+
+@dataclass
+class Event:
+    public_key: str = None
+    content: str = None
+    created_at: int = None
+    kind: int = EventKind.TEXT_NOTE
+    tags: List[List[str]] = None
+    id: str = None
+    signature: str = None
+
+
+    def __post_init__(self):
+        if self.content is not None and not isinstance(self.content, str):
+            # DMs initialize content to None but all other kinds should pass in a str
             raise TypeError("Argument 'content' must be of type str")
 
-        self.public_key = public_key
-        self.content = content
-        self.created_at = created_at or int(time.time())
-        self.kind = kind
-        self.tags = tags
-        self.signature = signature
-        self.id = id or Event.compute_id(self.public_key, self.created_at, self.kind, self.tags, self.content)
+        if self.created_at is None:
+            self.created_at = int(time.time())
+
+        # Can't initialize the nested type above w/out more complex factory, so doing it here
+        if self.tags is None:
+            self.tags = []
+
+        if self.id is None:
+            self.compute_id()
+
 
     @staticmethod
-    def serialize(public_key: str, created_at: int, kind: int, tags: "list[list[str]]", content: str) -> bytes:
+    def serialize(public_key: str, created_at: int, kind: int, tags: List[List[str]], content: str) -> bytes:
         data = [0, public_key, created_at, kind, tags, content]
         data_str = json.dumps(data, separators=(',', ':'), ensure_ascii=False)
         return data_str.encode()
 
-    @staticmethod
-    def compute_id(public_key: str, created_at: int, kind: int, tags: "list[list[str]]", content: str) -> str:
-        return sha256(Event.serialize(public_key, created_at, kind, tags, content)).hexdigest()
+
+    def compute_id(self):
+        self.id = sha256(Event.serialize(self.public_key, self.created_at, self.kind, self.tags, self.content)).hexdigest()
+
+
+    def add_pubkey_ref(self, pubkey:str):
+        """ Adds a reference to a pubkey as a 'p' tag """
+        self.tags.append(['p', pubkey])
+        self.compute_id()
```

Comment:
Because `compute_id` is called when signing, is this actually needed?

Reply:
I went back and forth on all these calls to `compute_id`. They are almost all redundant, but at the same time I didn't ever want the object to have an `id` that was out of sync with itself.

Actually... going to refactor `Event.id` to be a property and just recompute on the fly. Nothing should ever set the Event's `id`, only ask for it. Probably never made sense for `id` to be an input var on a new Event.

```
    @property
    def id(self) -> str:
        # Always recompute the id to reflect the up-to-date state of the Event
        return Event.compute_id(self.public_key, self.created_at, self.kind, self.tags, self.content)
```